### PR TITLE
`QueueAdapters` now does lazy lookup

### PR DESCRIPTION
**Pull request by tamird**
@rafaelfranca @jeremy

### CODE REVIEW

File: activejob/lib/active_job/queue_adapters.rb [hunk 1]
```diff
@@ -48,5 +48,32 @@ module QueueAdapters
...
+    class << self
+      def instance(name, *adapter_args, &block)
+        hash = ([name, block] << adapter_args).hash
+
+        return instances.fetch(hash) if instances.key?(hash)
+
+        mutex.synchronize do
+          return instances.fetch(hash) if instances.key?(hash)
+
+          @instances[hash] ||= lookup(name).new(*adapter_args, &block)
```
Comment (rafaelfranca) [inline activejob/lib/active_job/queue_adapters.rb]: Does we need to cache?

Comment (tamird) [inline activejob/lib/active_job/queue_adapters.rb]: I think so - in particular, we may have an adapter that carries a connection object on it - we would not want to create multiple identical connections

File: activejob/lib/active_job/queue_adapters.rb [hunk 2]
```diff
@@ -48,5 +48,32 @@ module QueueAdapters
...
+        end
+      end
+
+      private
+
+      attr_accessor :mutex, :instances
```
Comment (cristianbica) [inline activejob/lib/active_job/queue_adapters.rb]: private stuff should be indented by 2 spaces

File: activejob/lib/active_job/queue_adapters.rb [hunk 3]
```diff
@@ -48,5 +48,32 @@ module QueueAdapters
...
+    ADAPTER = 'Adapter'.freeze
+
+    class << self
+      def instance(name, *adapter_args, &block)
+        hash = ([name, block] << adapter_args).hash
+
+        return instances.fetch(hash) if instances.key?(hash)
+
+        mutex.synchronize do
+          return instances.fetch(hash) if instances.key?(hash)
```
Comment (cristianbica) [inline activejob/lib/active_job/queue_adapters.rb]: Is this like necessary? This line is accomplished by the `||=` below, right?

File: activejob/lib/active_job/queue_adapters.rb [hunk 4]
```diff
@@ -48,5 +48,32 @@ module QueueAdapters
     autoload :SneakersAdapter
     autoload :SuckerPunchAdapter
     autoload :TestAdapter
+
+    ADAPTER = 'Adapter'.freeze
+
+    class << self
+      def instance(name, *adapter_args, &block)
+        hash = ([name, block] << adapter_args).hash
```
Comment (thedarkone) [inline activejob/lib/active_job/queue_adapters.rb]: I don't think this is a correct use of `.hash` method, you are not guaranteed to get a different `hash` value for different `([name, block] << adapter_args)`. See also discussion in #15458.

Comment (tamird) [inline activejob/lib/active_job/queue_adapters.rb]: pretty sure it's a correct use - none of the mutable values are stored here.

Comment (thedarkone) [inline activejob/lib/active_job/queue_adapters.rb]: I think you are misunderstanding what I am worried about. How you read the discussion in #15458?

The problem is that this is allowed (though admittedly highly unlikely):
```ruby
['AdapterName', Proc.new {}, :arg1, :arg2].hash # => 12345
['DifferentAdapterName', Proc.new {}, :different_arg].hash # => 12345
```

Comment (tamird) [inline activejob/lib/active_job/queue_adapters.rb]: Oh, I skipped that part. Agree with @sgrif though - I think this argument is absurd.

Comment (sgrif) [inline activejob/lib/active_job/queue_adapters.rb]: I have changed my opinion on the matter, this is a valid argument

**Comment by thedarkone** [comment]
Correct me, if I am mistaken: isn't `QueueAdapter.interpret_adapter` not a hot path? Why are we rolling out a custom caching mechanism, why not simply use a plain mutex around the whole `#interpret_adapter` method?

Also, does setting `QueueAdapterClass.queue_adapter=` need to be thread-safe? The general rule of thumb in Rails is that class-level configuration assignments are on only to done during a single-threaded boot sequence?

Is there a reason you're not using `ThreadSafe::Cache` (ie: `cache = ThreadSafe::Cache.new; cache.compute_if_absent(some_key) {lookup(name).new(*adapter_args, &block)}`)?

**Comment by tamird** [comment]
> Correct me, if I am mistaken: isn't QueueAdapter.interpret_adapter not a hot path? Why are we rolling out a custom caching mechanism, why not simply use a plain mutex around the whole #interpret_adapter method?

Probably true - just erring on the side of caution.

> Also, does setting QueueAdapterClass.queue_adapter= need to be thread-safe? The general rule of thumb in Rails is that class-level configuration assignments are on only to done during a single-threaded boot sequence?

Again, probably true - just erring on the side of caution.

> Is there a reason you're not using ThreadSafe::Cache (ie: cache = ThreadSafe::Cache.new; cache.compute_if_absent(some_key) {lookup(name).new(*adapter_args, &block)})?

Just plain ignorance. I'll switch to that.

File: activejob/lib/active_job/queue_adapters.rb [hunk 5]
```diff
@@ -48,5 +48,25 @@ module QueueAdapters
...
+    class << self
+      def instance(name, *adapter_args, &block)
+        cache.compute_if_absent(([name, block] << adapter_args).hash) do
+          lookup(name).new(*adapter_args, &block)
+        end
+      end
+
+      private
+
+      attr_accessor :cache
```
Comment (jeremy) [inline activejob/lib/active_job/queue_adapters.rb]: We can use a `@cache` ivar, no need for private reader/writer methods

File: activejob/lib/active_job/queue_adapters.rb [hunk 6]
```diff
@@ -48,5 +48,25 @@ module QueueAdapters
...
     autoload :TestAdapter
+
+    ADAPTER = 'Adapter'.freeze
+
+    class << self
+      def instance(name, *adapter_args, &block)
+        cache.compute_if_absent(([name, block] << adapter_args).hash) do
+          lookup(name).new(*adapter_args, &block)
+        end
+      end
```
Comment (jeremy) [inline activejob/lib/active_job/queue_adapters.rb]: Caching is fine, but worth considering starting without it. The caller is asking for an instance of the named queue adapter — any instance.

Maybe we should dial this back to a simpler "here's an adapter name, give me its class" interaction to start. Then `#interpret_adapter` calls `ActiveJob::QueueAdapters.lookup(name_or_adapter_or_class)` (or `.fetch` or `#[]` or …) and calls `.new` itself.

We can tackle resource building, pooling, etc mechanics later when we have more pressure exerted on them.

File: activejob/lib/active_job/queue_adapters.rb (hunk 5 above)
Comment (tamird) [inline activejob/lib/active_job/queue_adapters.rb]: changed

File: activejob/lib/active_job/queue_adapters.rb (hunk 6 above)
Comment (tamird) [inline activejob/lib/active_job/queue_adapters.rb]: sure, that seems fine

**Comment by tamird** [comment]
Removed the caching - I'll save it for a future PR.

**Pull request merged**